WINDOWS: enable and fix failures in test_runtime_env_complicated

mattip opened the pull request:
Continuation of #22342 to enable runtime_env tests on windows. Part of #22186 

Enable runtime_env tests in test_runtime_env_complicated. The fixes involved getting paths correct in conda, virtualenvs, and base python, since there are differences between the layouts on linux and windows.

mattip commented on python/ray/_private/runtime_env/conda_utils.py:
```diff
@@ -192,6 +199,8 @@ def exec_cmd_stream_to_logger(
     The last n_lines lines of output are also returned (stdout and stderr).
     """
+    if "env" in kwargs and _WIN32 and "PATH" not in [x.upper() for x in kwargs.keys]:
+        raise ValueError("On windows, Popen requires 'PATH' in 'env'")
```
This change provides a nicer error message on windows when `env` is used but `PATH` is not provided. The default on windows is an empty `PATH`, which will almost always fail.

mattip commented on python/ray/_private/runtime_env/pip.py:
```diff
@@ -132,9 +137,8 @@ def _create_or_get_virtualenv(cls, path: str, cwd: str, logger: logging.Logger):
-            os.path.join(os.path.dirname(python), "..")
-        )
+
+        current_python_dir = sys.prefix
```
Use the canonical method to find the top-level python directory. On windows `python.exe` can be either in the top-level directory or in `Scripts`, so using `sys.prefix` works around complicated logic.

mattip replied:
moved this to windows-only to make sure it does not break non-windows

mattip commented on python/ray/_private/runtime_env/pip.py:
```diff
@@ -189,9 +193,7 @@ def _create_or_get_virtualenv(cls, path: str, cwd: str, logger: logging.Logger):
-        exit_code, output = exec_cmd_stream_to_logger(
-            create_venv_cmd, logger, cwd=cwd, env={}
-        )
+        exit_code, output = exec_cmd_stream_to_logger(create_venv_cmd, logger, cwd=cwd)
```
Remove the empty `env={}` which breaks windows (here and in other calls to `exec_cmd_stream_to_logger`)

mattip replied:
reverted this since it might be affecting linux

mattip commented:
This passes for me locally after properly setting up conda. CI is failing left and right, the relevant part seems to be the [windows run of test_runtime_env_complicated](https://buildkite.com/ray-project/ray-builders-pr/builds/24690#2d8c5bce-3742-4ce4-b8ab-0e5c92e79812/3161-5064) which complains
```
CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.
If using 'conda activate' from a batch script, change your
invocation to 'CALL conda.bat activate'.

To initialize your shell, run
conda init <SHELL_NAME>
```

mattip commented:
@architkulkarni thoughts? I think the failing tests do not involve the runtime environments and are flakey in general.

architkulkarni commented:
Thanks for your help with this!  The `test_runtime_env_complicated` windows failure looks relevant, right? https://buildkite.com/ray-project/ray-builders-pr/builds/25859#eef1e1ca-7889-4407-ae62-ea32be487530/3103-8900

I've restarted the test just to get more data points.

mattip commented:
Yes, of course it is relevant. I wonder how I missed that failure. Sorry, I will try to fix it.

architkulkarni commented:
No problem, I think we have too many other flaky tests which hide the real failures :(  Anyways, thanks and I'm happy to pair program on this again any time!

mattip commented:
Weird. The [windows test is failing](https://buildkite.com/ray-project/ray-builders-pr/builds/26041#83edbbb7-11b8-49a0-8503-2783768fc231) but are marked passing in the summary here.

architkulkarni commented:
Odd, could it be that windows Build & test buildkite job hasn't been set up to show up as a separate check mark in github, so it only shows up at the top level `buildkite/ray-builders-pr ` check mark?

mattip commented:
I enabled more tests from `test_runtime_env_*.py` and skipped a few tests that are still failing in `runtime_env_complicated.py`, in an effort to get this PR to a point where it can be reviewed.

architkulkarni commented:
Sounds good!  I just reviewed it and it looks great.  I guess we should hold off on merging anything that removes the `NotImplementedError` from conda/pip on Windows until a suitable fraction of the tests are passing.  Or we can decide how to best signal it to users (maybe replacing `NotImplementedError` with a warning would be sufficient?)

mattip commented:
Yes, we should have more tests passing. It may take a few iterations to get CI to be happy with just these changes. I tried a few different scenarios and the tests seemed to pass locally: running with only conda init, running with conda activated, running with `RAY_CONDA_HOME` set. CI sometimes has a mind of its own, so let's see how this shakes out

mattip commented:
The calls to `await PipProcessor(target_dir, runtime_env, logger)` seem to be timing out. @architkulkarni do you know where any such timeout might be?

mattip commented:
I am a bit at a loss here. Is there an easy way to get a summary of which tests are failing?

mattip commented:
Also it seems `test_cli` on linux [is failing](https://buildkite.com/ray-project/ray-builders-pr/builds/26388#f64756e5-effc-47a0-9121-de06ffe9d624/1821-4766) with 
> 2022-03-10 06:12:50,053	WARNING services.py:2022 -- WARNING: The object store is using /tmp instead of /dev/shm because /dev/shm has only 2684354560 bytes available. This will harm performance! You may be able to free up space by deleting files in /dev/shm. If you are inside a Docker container, you can increase /dev/shm size by passing '--shm-size=4.01gb' to 'docker run' (or add it to the run_options list in a Ray cluster config). Make sure to set this to more than 30% of available RAM.

Is it passing on HEAD? Is this because of a random combination of tests leaving too much cruft on `/dev/shm`?

architkulkarni commented:
`serve:test_cli` indeed seems to be flaky: https://flaky-tests.ray.io/

I'm not sure about the timeout in PipProcessor, is it passing locally?

mattip commented:
PipProcessor passes locally.

kfstorm commented:
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

mattip commented:
@architkulkarni any ideas why the `test_requirements_files` test is reporting [ModuleNotFoundError: No module named 'redis'](https://buildkite.com/ray-project/ray-builders-pr/builds/27194#0f92309d-a313-4391-8249-ab904e0c581d/3023-6353)

mattip commented:
>  We can decide how to best signal it to users (maybe replacing `NotImplementedError` with a warning would be sufficient?)

Rebased and added a warning instead of the Error.

architkulkarni commented:
> @architkulkarni any ideas why the `test_requirements_files` test is reporting [ModuleNotFoundError: No module named 'redis'](https://buildkite.com/ray-project/ray-builders-pr/builds/27194#0f92309d-a313-4391-8249-ab904e0c581d/3023-6353)

Is the failure only happening in CI, or is it failing locally too?

I wonder if this is related? https://github.com/ray-project/ray/pull/22979/files It removes `redis` from the dependencies.  That PR passed all Windows tests so `import ray` should work... I'm not sure why `import ray` is failing in the conda environment.  @iycheng I know you don't have much context here, but let me know if you have any ideas about this off the top of your head.

mattip commented:
I wonder if the runtime environment is getting mixed up somehow between code from an older version with an environment from a newer version. But then I would expect to see it on linux as well.

architkulkarni commented:
> I wonder if the runtime environment is getting mixed up somehow between code from an older version with an environment from a newer version. But then I would expect to see it on linux as well.

So there is one difference here: For conda, we automatically inject the appropriate Ray version into the conda environment so the user doesn't have to specify it.  For released Ray versions and nightly wheels, we insert the wheel URL from AWS storage.  But for premerge CI, we need a workaround because we don't release and upload wheels for each PR.

For Linux, we have a pipeline that build a wheel in CI and then passes in the path of that wheel to the runtime env test.  We only do this in CI, and we do it by specifying the environment variables `"CI"` and `"RAY_CI_POST_WHEEL_TESTS"`: https://github.com/ray-project/ray/blob/56287d63e5d650327dff4d06fe6561746ac6583a/python/ray/_private/runtime_env/conda.py#L114
https://github.com/ray-project/ray/blob/6eb805b357754a05d41dbf145febd132490a0a8f/.buildkite/pipeline.yml#L28

For Windows, we don't have this set up, so I'm guessing it's using the "local dev mode", which is used by developers who are working with the source and uses a different mechanism to link to the Python site without building a wheel: https://github.com/ray-project/ray/blob/56287d63e5d650327dff4d06fe6561746ac6583a/python/ray/_private/runtime_env/conda.py#L62 This is triggered using a different environment variable flag: https://github.com/ray-project/ray/blob/39a6c054d385d162d71c295f6d1abe30322dcbd2/python/ray/tests/test_runtime_env_complicated.py#L37

I would expect both of these to work, though.  I'm not sure what "older version" it could be pulling from in CI, but that does sound like a reasonable hypothesis.

mattip commented:
I defined CI=1 and RAY_CI_POST_WHEEL_TESTS=1 for windows. Let's see if that helps

architkulkarni commented:
Sounds good, thanks! In the Windows pipeline is there a wheel being built at the desired location? Something corresponding to this line in the linux pipeline? https://github.com/ray-project/ray/blob/6eb805b357754a05d41dbf145febd132490a0a8f/.buildkite/pipeline.yml#L25

This part of the code describes the expected location for the wheel:
 https://github.com/ray-project/ray/blob/56287d63e5d650327dff4d06fe6561746ac6583a/python/ray/_private/runtime_env/conda.py#L114-L120

architkulkarni commented:
I tried something like that here in case it's helpful, not sure if it worked though: https://github.com/ray-project/ray/pull/22186/commits/1377f199fbeedeb1110963ec7f08d6c96f6c3110

mattip commented:
> I tried something like that here in case it's helpful, not sure if it worked though: https://github.com/ray-project/ray/commit/1377f199fbeedeb1110963ec7f08d6c96f6c3110

Could you give a little more context to that diff? Is it part of a PR or a fork?

architkulkarni commented:
> > I tried something like that here in case it's helpful, not sure if it worked though: [1377f19](https://github.com/ray-project/ray/commit/1377f199fbeedeb1110963ec7f08d6c96f6c3110)
> 
> Could you give a little more context to that diff? Is it part of a PR or a fork?

Ah sorry!  It's from https://github.com/ray-project/ray/pull/22186, not sure why that didn't show up in the commit

mattip commented:
Linux is failing `test_basic_2_client_mode` where `test_actor_recursive` [times out](https://buildkite.com/ray-project/ray-builders-pr/builds/27289#b115bcc2-df40-4716-b8eb-fd673b3f1d52/117-227). I see that test is flaky on linux as well as windows. Also there is an error when [fetching a repository](https://buildkite.com/ray-project/ray-builders-pr/builds/27289#8a8388b0-609c-4b18-a261-7c3281e94d0c/960-1027). Neither of those seem to be connected to this PR.

mattip commented:
I cherrypicked the reporter_agent fix for missing USERPROFILE from PR #23117.

mattip commented:
There are two tests failing:
- `test_experimental_package` in `test_runtime_env_complicated.py` [cannot find the yaml file](https://buildkite.com/ray-project/ray-builders-pr/builds/27429#c8b15130-111f-4c70-9e8f-b6de05970a83/2979-4326), which seems to be some problem with UNC paths (they begin with `\\?\C:`
- `test_install_failure_logging` in `test_runtime_env_conda_and_pip_2.py` fails about 1/2 of the cases. 

For now I will skip these two.

architkulkarni commented:
> There are two tests failing:
> 
> * `test_experimental_package` in `test_runtime_env_complicated.py` [cannot find the yaml file](https://buildkite.com/ray-project/ray-builders-pr/builds/27429#c8b15130-111f-4c70-9e8f-b6de05970a83/2979-4326), which seems to be some problem with UNC paths (they begin with `\\?\C:`
> * `test_install_failure_logging` in `test_runtime_env_conda_and_pip_2.py` fails about 1/2 of the cases.
> 
> For now I will skip these two.

That sounds good to me.  In the future I think we can get rid of the `experimental_package` code, since it's an early experimental version of runtime_env that's been superseded by the current runtime env implementation.

mattip commented:
rebased to get the split of `test_runtime_env.py` from master

mattip commented:
@architkulkarni All windows CI is passing.

mattip commented on dashboard/modules/reporter/reporter_agent.py:
```diff
@@ -275,7 +275,10 @@ def _get_disk_usage():
+        if sys.platform == "win32":
+            root = psutil.disk_partitions()[0].mountpoint
+        else:
+            root = os.sep
```
Taken from #23117

mattip commented on python/ray/serve/tests/test_pipeline_driver.py:
```diff
@@ -15,6 +15,7 @@ def my_resolver(a: int):
     return a
 
 
+@pytest.mark.skipif(sys.platform == "win32", reason="Fails on windows.")
```
This whole feature is reverted in PR #23358. It was added quite recently. For now skip the failing test

mattip commented on python/ray/tests/BUILD:
```diff
@@ -65,6 +65,7 @@ py_test_module_list(
 py_test_module_list(
   files = [
     "test_client.py",
+    "test_client_reconnect.py",
```
This test timesout on windows, so move it to large from medium

mattip commented on python/ray/tests/test_runtime_env.py:
```diff
@@ -276,7 +270,7 @@ def f():
 
 
-@pytest.mark.skipif(sys.platform == "win32", reason="pip not supported on Windows.")
+@pytest.mark.skipif(sys.platform == "win32", reason="Hangs on windows.")
```
Investigate this later, the PR is large enough as-is.

mattip commented on python/ray/tests/test_runtime_env_complicated.py:
```diff
@@ -527,7 +568,7 @@ def test_experimental_package_lazy(shutdown_only):
 
 
-@skipIf(sys.platform == "win32", "Fail to create temp dir.")
+@pytest.mark.skipif(_WIN32, reason="requires tar cli command")
```
Refactor the `tar` requirement later, this PR is big enough as-is.

mattip commented on src/ray/common/ray_config_def.h:
```diff
@@ -351,7 +351,12 @@ RAY_CONFIG(int64_t, task_rpc_inlined_bytes_limit, 10 * 1024 * 1024)
+RAY_CONFIG(uint32_t, agent_register_timeout_ms, 100 * 1000)
+#else
 RAY_CONFIG(uint32_t, agent_register_timeout_ms, 30 * 1000)
+#endif
```
As discussed elsewhere, this is needed since process startup time is much longer on windows

mattip commented on ci/travis/ci.sh:
```diff
@@ -179,6 +178,10 @@ test_python() {
+      --test_env=USERPROFILE="${USERPROFILE}" \
+      --test_env=CI=1 \
+      --test_env=RAY_CI_POST_WHEEL_TESTS=1 \
+      --test_output=streamed \
```
Streaming the output forced tests to not be run in parallel, but it does not seem to increase the time required for CI. Part of PR #23188, repeated here.

architkulkarni reviewed (approved):
Looks good to me, thanks for the hard work!  What do you think the docs should say about this? Right now it says https://github.com/ray-project/ray/blob/master/doc/source/ray-core/handling-dependencies.rst#runtime-environments "currently only supported on macOS and Linux."

It looks like after this PR we can say that the `working_dir`, `env_vars`, `pip` and `conda` fields are experimentally supported, but perhaps not `py_modules` and `container`?

mattip commented:
I rebased to get #23373 and also changed the documentation.

architkulkarni commented on doc/source/ray-core/handling-dependencies.rst:
```diff
@@ -52,7 +52,7 @@ Runtime environments
 .. note::
 
-    This feature requires a full installation of Ray using ``pip install "ray[default]"``. This feature is available starting with Ray 1.4.0 and is currently only supported on macOS and Linux.
+    This feature requires a full installation of Ray using ``pip install "ray[default]"``. This feature is available starting with Ray 1.4.0 and is currently supported on macOS and Linux. It is experimentally supported on Widows.
```
```suggestion
    This feature requires a full installation of Ray using ``pip install "ray[default]"``. This feature is available starting with Ray 1.4.0 and is currently supported on macOS and Linux. It is experimentally supported on Windows.
```

mattip commented:
The bad news is that one of the windows shards is [failing](https://buildkite.com/ray-project/ray-builders-pr/builds/27822#e64c0204-b7b2-4bd4-812a-e8f2d50bac26/2835-8134). The less bad news is that it is in a test tht seems to be unrelated to this PR, in `test_worker_capping`. I could not see any record of the test actually being collected and run, so this seems like a separate issue.

mattip commented:
The `setting-up-bootstrap-env-ml-dock` failure is due to [a download failure](https://buildkite.com/ray-project/ray-builders-pr/builds/27822#1d0634ba-772d-4660-96d5-7562aab8f734), apparently not connected to this PR.

mattip commented:
That failure is gone now there is a different one, again, not on windows: `two_step_game_maddpg` from `//rllib:examples/recommender_system_with_recsim_and_slateq_tf2`. It seems windows failed `test_worker_stats` in `//python/ray/tests:test_metrics` but the shard was re-run and it passed.

architkulkarni commented:
Tests look good!  @pcmoritz shall we merge this?  I don't have merge privileges.
